### PR TITLE
Add L0_libtorch_io_names test and update docs for new Naming Convention for PyTorch backend

### PR DESCRIPTION
**Pull request by CoderHam**
- Tests fall back to enforcing ordering for cases when index and model arguement naming convention for I/O is not met
- These models that would earlier fail to load, now load succesfully and the input/outputs assume the order they are specified in inside the model configuration file

### CODE REVIEW

File: qa/L0_libtorch_io_names/io_names_client.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        io_names = ["INPUT0", "INPUT1", "OUTPUT__0", "OUTPUT__1"]
+        self._infer_helper("libtorch_output_index", io_names)
+
+    def test_no_output_index(self):
+        io_names = ["INPUT0", "INPUT1", "OUTPUT0", "OUTPUT1"]
+        self._infer_helper("libtorch_no_output_index", io_names)
+
+    def test_no_arguements_no_output_index(self):
+        io_names = ["INPUTA", "INPUTB", "OUTPUTA", "OUTPUTB"]
+        self._infer_helper("libtorch_no_arguements_output_index", io_names)
```
Comment (GuanLuo) [inline qa/L0_libtorch_io_names/io_names_client.py:92]: Add test for the mix of different rules.

File: qa/L0_libtorch_io_names/io_names_client.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
```
Comment (dyastremsky) [inline qa/L0_libtorch_io_names/io_names_client.py:93]: Potential tests to add:
Test when one of inputs is valid argument and has index and the other does not (Can also do this with outputs; clarify/test intended behavior)
Test where one of two conditions is met (argument or index) to clarify/test intended behavior.

Comment (CoderHam) [reply qa/L0_libtorch_io_names/io_names_client.py:93]: Thanks for the suggestions. Added these in

File: qa/L0_libtorch_io_names/io_names_client.py (hunk 1 above)
Comment (CoderHam) [reply qa/L0_libtorch_io_names/io_names_client.py:92]: Added the mix of rules as well

1. ["INPUTA", "INPUT__1", "OUTPUTA", "OUTPUT__1"]
2. ["INPUT0", "INPUTB", "OUTPUTA", "OUTPUT__1"]
3. ["INPUT0", "INPUT__1", "OUTPUT0", "OUTPUT__1"]
4.  ["INPUT1", "INPUT0", "OUTPUT__1", "OUTPUT__0"] (unordered)

**Review by dyastremsky — DISMISSED** [review]

**Review by dyastremsky — APPROVED** [review]

File: qa/L0_libtorch_io_names/io_names_client.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    def test_mix_index(self):
+        io_names = ["INPUTA", "INPUT__1", "OUTPUTA", "OUTPUT__1"]
```
Comment (GuanLuo) [inline qa/L0_libtorch_io_names/io_names_client.py:100]: Should we fail the model loading if we see only a subset inputs have index suffix? (same for output) If user specify index suffix in config, the intention is clear that the user is aware of the naming convention and would like to make sure the ordering is as intended. Say he specify the following by mistake: `"INPUT1", "INPUT__0"`, with the intended ordering to be `2nd model input, 1st model input`. It appears that the backend will treat this as config ordering and make it `1st model input, 2nd model input`

Comment (CoderHam) [reply qa/L0_libtorch_io_names/io_names_client.py:100]: We do not fail. We allow users to specify model names and log the warning that we enforce strict ordering from the model config. This holds prior to my current changes i.e. users can now specify INPUT__0 twice in the model config and that model would be loaded successfully.

**Review by dyastremsky — DISMISSED** [review]

**Review by dyastremsky — DISMISSED** [review]

**Review by tanmayv25 — DISMISSED** [review]

**Comment by CoderHam** [comment]
@tanmayv25 @dyastremsky I updates the docs to be more clear about the config naming convention for I/O in the PyTorch backend

File: docs/model_configuration.md [hunk 4]
```diff
@@ -138,37 +138,47 @@ expected by the model.
...
-model that expects the input of the form:
+For example, if the forward function for the Torchscript model was defined as
+`forward(self, input0, input1)`, the first and second inputs should be named
+"input0" and "input1" respectively. 
+
+2. `<name>__<index>`: Where \<name\> can be any string and \<index\> is an
+integer index that refers to the position of the corresponding input/output.
+
+This means that if there are two inputs and two outputs, the first and second
+inputs would could named "INPUT__0" and "INPUT__1" and the first and second
```
Comment (dyastremsky) [inline docs/model_configuration.md:159]: "would could" duplicate word? Should also match the below line, I believe. I.e. they both can or must be named that.

File: docs/model_configuration.md [hunk 5]
```diff
@@ -138,37 +138,47 @@ expected by the model.
...
+Due to the absence of sufficient metadata for inputs/outputs in TorchScript
+model files, the "name" attribute of inputs/outputs in the configuration must
+follow specific naming conventions. These are detailed below.
 
-Similarly, for the inputs, the first and second inputs would be named "INPUT__0"
-and "INPUT__1" respectively. Alternatively, if the forward function for the
-Torchscript model was defined as `forward(self, input0, input1)`, the first and
-second inputs could also be named "input0" and "input1" respectively. 
+1. [Only for Inputs] When the input is not a Dictionary of Tensors, the input
+names in the configuration file can mirror the names of the input arguements to
```
Comment (dyastremsky) [inline docs/model_configuration.md:148]: Arguments, typo.

Comment (dyastremsky) [reply docs/model_configuration.md:148]: Also, mixing of can and should. We need to be consistent about whether a user can do something or should/must do it.

File: docs/model_configuration.md [hunk 6]
```diff
@@ -138,37 +138,47 @@ expected by the model.
...
 ```
 {'A': tensor1, 'B': tensor2}
 ```
 
-Then the input fields in the configuration do not need to follow the "\<name\>__\<index\>" 
-convention. Instead, the names of the inputs in this case must map to the string
-value 'key' for that specific tensor. In this case the inputs would be "A" and "B",
-where input "A" refers to value corresponding to tensor1 and "B" refers to the
-value corresponding to tensor2.
+The input names in the configuration in this case must not to follow the above
```
Comment (dyastremsky) [inline docs/model_configuration.md:177]: "Must not follow" (remove to)

**Review by dyastremsky — CHANGES_REQUESTED** [review]

**Review by dyastremsky — APPROVED** [review]

**Review by tanmayv25 — APPROVED** [review]

**Pull request merged**